Tigera calico policy for arc enabled eks/gke (#950)
microsoft/azure_arc: docs/azure_arc_jumpstart/azure_arc_k8s/day2/calico/_index.md

---
type: docs
title: "Use the Azure Policy Add-on to audit Calicocloud / calico enterprise"
linkTitle: "Use the Azure Policy Add-on to create a custom Azure Policy for auditing calicocloud/calico enterprise components"
weight: 4
description: >
---

## Use the Azure Policy Add-on with a custom Azure Policy to verify the calicocloud/calico enterprise components

> The following README will guide you on how to use a custom Azure Policy [Azure Policy for Kubernetes](https://docs.microsoft.com/en-us/azure/governance/policy/concepts/policy-for-kubernetes#:~:text=Azure%20Policy%20extends%20Gatekeeper%20v3,Kubernetes%20clusters%20from%20one%20place.) on an Azure Arc connected Kubernetes Service cluster to verify whether Calico network policy is enabled or not.

Reviewer: ```suggestion
The following README will guide you on how to use a custom Azure Policy [Azure Policy for Kubernetes](https://docs.microsoft.com/en-us/azure/governance/policy/concepts/policy-for-kubernetes#:~:text=Azure%20Policy%20extends%20Gatekeeper%20v3,Kubernetes%20clusters%20from%20one%20place.) on an Azure Arc-connected Kubernetes cluster to verify whether [Calico network policy](https://projectcalico.docs.tigera.io/about/about-network-policy) is enabled or not.
```

> > **Note: This guide assumes you already deployed an EKS/GKE cluster and connected it to Azure Arc. If you haven't, this repository offers you a way to do so in an automated fashion using [Terraform for eks](https://azurearcjumpstart.io/azure_arc_jumpstart/azure_arc_k8s/eks/eks_terraform/) or [Terraform for gke](https://azurearcjumpstart.io/azure_arc_jumpstart/azure_arc_k8s/gke/gke_terraform/).**

Reviewer: ```suggestion
> **Note: This guide assumes you already deployed an Amazon Elastic Kubernetes Service (EKS) or Google Kubernetes Engine (GKE) cluster and connected it to Azure Arc. If you haven't, this repository offers you a way to do so in an automated fashion using these couple of Jumpstart scenarios:
- [Deploy EKS cluster and connect it to Azure Arc using Terraform](https://azurearcjumpstart.io/azure_arc_jumpstart/azure_arc_k8s/eks/eks_terraform/)
- [Deploy GKE cluster and connect it to Azure Arc using Terraform](https://azurearcjumpstart.io/azure_arc_jumpstart/azure_arc_k8s/gke/gke_terraform/).**
```

## Prerequisites

* Clone the Azure Arc Jumpstart repository

    ```shell
    git clone https://github.com/microsoft/azure_arc.git
    ```

* [Install or update Azure CLI to version 2.25.0 and above](https://docs.microsoft.com/en-us/cli/azure/install-azure-cli?view=azure-cli-latest). Use the below command to check your current installed version.

  ```shell
  az --version
  ```

* As mentioned, this guide starts at the point where you already have a connected EKS/GKE cluster to Azure Arc.

    ![Existing EKS Azure Arc connected cluster](./arcdemo-eks.png)

>     ![Existing GKE Azure Arc connected cluster](./arcdemo-gke.png)

Reviewer: Images needs to be numberd

Author: chang their name to 15 & 16

> * Before installing the Azure Policy Add-on or enabling any of the service features, your subscription must enable the Microsoft.PolicyInsights resource provider and create a role assignment for the cluster service principal. To do that, open [Azure Cloud Shell](https://shell.azure.com/) and run either the Azure CLI or Azure PowerShell command.

Reviewer: ```suggestion
* Before installing the Azure Policy Add-on or enabling any of the service features, your subscription must enable the _Microsoft.PolicyInsights_ resource provider and create a role assignment for the cluster service principal. To do that, open [Azure Cloud Shell](https://shell.azure.com/) and run either the Azure CLI or PowerShell command.
```

    ![Open Azure Cloud Shell](./03.png)

    Azure CLI:

    ```shell
    az provider register --namespace 'Microsoft.PolicyInsights'
    ```

>     Azure PowerShell:

Reviewer: ```suggestion
    PowerShell:
```

    ```powershell
    Register-AzResourceProvider -ProviderNamespace 'Microsoft.PolicyInsights'
    ```

>     To verify successful registration, run either the below Azure CLI or Azure PowerShell command.

Reviewer: ```suggestion
    To verify successful registration, run either the below Azure CLI or PowerShell command.
```

    Azure CLI:

    ```shell
    az provider show --namespace 'Microsoft.PolicyInsights'
    ```

>     Azure PowerShell:

Reviewer: ```suggestion
    PowerShell:
```

    ```powershell
    Get-AzResourceProvider -ProviderNamespace 'Microsoft.PolicyInsights'
    ```

    ![AzResourceProvider Bash](./04.png)

> * Azure Policy installed in Arc Connected Cluster

Reviewer: ```suggestion
* Azure Policy installed in Azure Arc-Connected Cluster
```

 > by running the the ```kubectl get pods -n gatekeeper-system ``` command, you will notice all pods have been deployed.

> ![Showing pods deployment](./05.png)

Reviewer: Image has a transparent background which makes it confusing to read

Author: Change the pic without any transpanrency

> ## Deploy a customized Azure policy to verify if calicocloud/calico enterprise components enabled in your Azure Arc Kubernetes cluster.

Reviewer: ```suggestion
## Deploy a customized Azure policy to verify if calicocloud/calico enterprise components are enabled in your Azure Arc-enabled Kubernetes cluster.
```

Reviewer: "calicocloud/calico enterprise"

Shouldn't this be "Calico Cloud or Calico Enterprise"?

Author: You are right, revise my words

> * In the Search bar, look for *Policy* and click on *Definitions* which will show you all of the available Azure policies.

Reviewer: ```suggestion
* In the Search bar, look for _Policy_ and click on _Definitions_ which will show you all of the available Azure policies.
```

    ![Searching for Azure Policy definitions](./06.png)

>     ![Searching for Azure Policy definitions](./07.png)

Author: add red arrow

> * Click on *+ Policy definition* to add a new policy. 

Reviewer: ```suggestion
* Click on _+ Policy definition_ to add a new policy. 
```

>     ![Create a Azure policy for Kubernetes cluster](./08.png)

Reviewer: Add red box around "Policy definition"

    ![Create a Azure policy for Kubernetes cluster](./09.png)

> * In the above example, the *Definition location* of the policy represent the Azure subscription. Also, make sure *Category* is set to *Kubernetes*. We will use `audit` as default effect for this policy. 

Reviewer: ```suggestion
* In the above example, the _Definition location_ of the policy represents the Azure subscription. Also, make sure _Category_ is set to _Kubernetes_. We will use `audit` as the default effect for this policy. 
```

Reviewer: This sentence should come before 10.png IMO (if so, please update to "In the below example"

Author: move the sentence before 09.png

>   You can either copy the `calicoapipolicy.json` file located in this folder into `POLICY RULE` or create a policy directly with azure cli with below command.

Reviewer: ```suggestion
  You can either copy the `calicoapipolicy.json` file located in this folder into `POLICY RULE` or create a policy directly with Azure CLI with the below command.
```

  ```shell
  az policy definition create --name 'arcdemo-calicoapi-policy' --display-name 'Audit Calico api server' --description 'This policy audit that if calicocloud/calico enterprise components exist' --rules 'calicoapipolicy.json' --mode All
  ```

> * Once the policy is been created, you should be able to find it in *Policy | Definitions* 

Reviewer: ```suggestion
* Once the policy is been created, you should be able to find it in _Policy | Definitions_.
```

Author: These two changes been removed since we won't create customer policy

>   ![Calico Azure policy ](./10.png) need change

Reviewer: ```suggestion
  ![Calico Azure policy ](./10.png)
```

> * Click the policy you created, and assign it to your resource group which include the arc clusters. Alternatively, you can assign the policy to entre Azure subscription.

Reviewer: ```suggestion
* Click the policy you created, and assign it to your Azure resource group which includes the Azure Arc-enabled Kubernetes clusters. Alternatively, you can assign the policy to entre Azure subscription.
```

Author: manually revise this sentence to "* Click this policy, and assign it to your Azure resource group which includes the Azure Arc-enabled Kubernetes clusters. Alternatively, you can assign the policy to entre Azure subscription."

>   ![Assign Calico Azure policy ](./11.png) need change

Reviewer: ```suggestion
  ![Assign Calico Azure policy ](./11.png)
```

  

* After the assignment, the policy task will start the evaluation against arc enabled cluster under your resource group. If you have 2 clusters installed calicocloud/calico enterprise, it will show as  "compliant" with `2 out of 2 ` (The gatekeeper will look for deployment with label 'apiserver' as audit labellist). To check this, go back to the main Policy page in the Azure portal.

    > **Note: The process of evaluation against the cluster can take 30min.**

>     ![Azure policy evaluation](./12.png)

Reviewer: This does not look like the right image as it shows AKS. Also, Resource names can not have a personal reference, they need to be natural.

Author: replace another image with gke arc

>     > If you create ks8 without enabling azure policy, then the audit result of this policy will be `0 out of 0` as `Compliant`

Reviewer: This feels a bit "missing" and also confusing. Need better explanation

Author: add more explanation

## Clean up environment

Complete the following steps to clean up your environment.

> * From the Policy page in the arc eks/gke portal, disable the extension of Azure policy under "Onboard to Azure Policy for Azure Arc enabled Kubernetes clusters​"

Reviewer: eks/gke portal?

Author: change the words

    ![Disable Azure Policy addon](./13.png)
    ![Disable Azure Policy addon](./14.png)

> * Delete the cluster as described in the [Terraform teardown instructions for eks](https://azurearcjumpstart.io/azure_arc_jumpstart/azure_arc_k8s/eks/eks_terraform/) or [Terraform teardown instructions for gke](https://azurearcjumpstart.io/azure_arc_jumpstart/azure_arc_k8s/gke/gke_terraform/) .

Reviewer: Please refer to the full Jumpstart scenarios names as described in my comment at the beginning of the README.

Author: changed
